static/panes/tool.ts

```diff
@@ -0,0 +1,676 @@
+// Copyright (c) 2018, Compiler Explorer Authors
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import _ from 'underscore';
+import {FontScale} from '../widgets/fontscale';
+import {Filter as AnsiToHtml} from '../ansi-to-html';
+import {Toggles} from '../widgets/toggles';
+import {ga} from '../analytics';
+import * as Components from '../components';
+import * as monaco from 'monaco-editor';
+import * as monacoConfig from '../monaco-config';
+import {options as ceoptions} from '../options';
+import * as utils from '../utils';
+import {PaneRenaming} from '../widgets/pane-renaming';
+import {saveAs} from 'file-saver';
+import {MonacoPane} from './pane';
+import {Hub} from '../hub';
+import {Container} from 'golden-layout';
+import {MonacoPaneState} from './pane.interfaces';
+import {SiteSettings} from '../settings';
+import {ComponentConfig, PopulatedToolInputViewState} from '../components.interfaces';
+import {ToolState} from './tool.interfaces';
+import {CompilerInfo} from '../../types/compiler.interfaces';
+import {CompilationResult} from '../../types/compilation/compilation.interfaces';
+import {ToolInfo} from '../../lib/tooling/base-tool.interface';
+import {MessageWithLocation} from '../../types/resultline/resultline.interfaces';
+
+function makeAnsiToHtml(color?: string): AnsiToHtml {
+    return new AnsiToHtml({
+        fg: color ? color : '#333',
+        bg: '#f5f5f5',
+        stream: true,
+        escapeXML: true,
+    });
+}
+
+export class Tool extends MonacoPane<monaco.editor.IStandaloneCodeEditor, any> {
+    private hub: Hub;
+    private editorContentRoot: JQuery<HTMLElementTagNameMap[keyof HTMLElementTagNameMap]>;
```

Comment:
Might be out of the loop, but could you elaborate on why this type is used over JQuery<HTMLElement> / JQuery<that specific element type>?

Reply:
In the case that we want to use this type, we should make it a local type so it's not repeated 20 times in this file

Reply:
This was auto-generated by my IDE, I'll just replace it